Tutorial ViT wave function

llviteritti opened the pull request:
We added a tutorial on the Vision Transformer (ViT) wave function.
First, we describe step-by-step the building blocks of the architecture. 
Then, we show how to train it for approximating the ground state of the J1-J2 Heisenberg model on a 10x10 square lattice.
At the end, we load a pretrained Transformer from Hugging Face in Netket.

We add the tutorial in the netket/docs/index.md.

CC : @riccardo-rende

gcarleo reviewed (approved):
thank you this is excellent, and long needed! please just remove those random warning messages otherwise it's good to go for me

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<p>Please remove the last line. This is on netket website. </p><p>just mention that you build the network using flax's legacy linen interface.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

llviteritti replied:
<p>Done</p>

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #13.    from flax.training import checkpoints
</pre><p>This is only needed to load from hugging face, right? </p><p>Could you move the pretrained-loading stuff down at the bottom, in the relevant section?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

llviteritti replied:
<p>Yes, we moved this line in the section <em>Pretrained model</em> and we removed <code>flax.config.update('flax_use_orbax_checkpointing', False)</code></p>

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #14.    flax.config.update('flax_use_orbax_checkpointing', False)
</pre><p>is this needed for the notebook?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

llviteritti replied:
<p>No, we removed it</p>

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<p>1.1. explicitly mention that the patches are of size b (in 1D) or b^2 in 2D</p><p>2 mention that the encoder is real-valued.</p><p>3.1 : explicitly mention that z is an element of R^d</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

llviteritti replied:
<ol><li>We discussed that a specific function for each geometry should be implemented to extract the patches from the spin configuration.</li><li>Done</li><li>Done</li></ol>

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<p>Please explain why in 2D the reshaping must be done like that. I've seen several people not get it.</p><p>To this end, you could draw <code>graph.draw()</code>  for a 4x4 square lattice and show that the plaquettes should have the sites [1,2,5,6] , [3,4,7,8] etc and that the way to get this is to do the reshaping as you do below.</p><p> </p><p>To explain this properly it might take an extra cell or2 </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

llviteritti replied:
<p>We replied to this point in the previous answer.</p>

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #5.        x = x.transpose(0, 1, 3, 2, 4)             # [L_eff, L_eff, b, b]
</pre><p>As you're dragging in einops, you could even use <code>einops.rearrange(x, "b</code>Ly by Lx bx -&gt;b Ly Lx by bx") which is more clear.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #11.        d_model : int # dimensionality of the embedding space
</pre><p>mention that this is the d of the equations above.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

llviteritti replied:
Done

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #15.            self.embed = nn.Dense(self.d_model, kernel_init=nn.initializers.xavier_uniform(), param_dtype=self.param_dtype)
</pre><p> </p><p>Add a <code>self.param_dtype</code> field please, just to educate people that it is param dtype and not dtype that should be set.</p><p>(Do not set dtype!) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

llviteritti replied:
Done

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<p>one equation does not render correctly in here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

llviteritti replied:
<p>In readthedocs works</p>

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #2.        L_eff: int   # lenght of the input sequence
</pre><p>Please remove L_eff, and infer it from the input with @compact. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

llviteritti replied:
<p>We think that it is more clear to have the setup method and don't mix it with @compact</p>

PhilipVinc commented on docs/tutorials/ViT-wave-function.ipynb:
```diff
@@ -0,0 +1,3423 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #5.        L_eff: int      # lenght of the input sequence
</pre><p>Please drop this as well, as it's inferred from the input.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/2004/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

llviteritti replied:
See previous answer

llviteritti commented:
We addressed all the the comments

gcarleo commented:
@PhilipVinc if there are no further comments I will merge this

PhilipVinc reviewed (dismissed):
Last week I had given a few comments in person to @llviteritti that I'd like addressed before merging. The important ones that come to my mind are:
 - Rename `L_eff` to something else, for example `N_patches`, as the current naming is confusing in 2D;
 - Rename `b` to `patch_size` or something similar. 
 - Rename `h` to `n_heads` or something similar.
 - The current version of the code leaves the freedom to specify independently the input size, number of patches and patch size. However only 2 of those quantities can effectively be 'set' by the user. This makes the code harder to understand to someone looking it for the first time. 

I would like that the entry-point, that is the ViT class, only provides 2 of those 3 things to be specified. My personal taste would be to specify `N` the input size and the patch size, and infer the number of patches from that. Also add an assert to the beginning of the `ViT.__call__` method to check that the input has the same number `N` of inputs as the ones specified in the constructor.

Essentially, what I propose is to change the example to be something like
```python
class ViT(nn.Module):
    N: int                  # input size
    patch_size: int  # linear patch size (so in 2D the actual size is patch_size**2)
    num_layers: int # number of layers
    d_model : int   # dimensionality of the embedding space
    n_heads: int          # number of heads
    transl_invariant: bool = False

    @property
    @nn.nowrap
    def n_patches(self):
         return self.N//self.patch_size

    def __call__(self, x):
           assert x.shape[-1] == self.N, "Size mismatch"
           ...
```

gcarleo commented:
ok thank you @PhilipVinc. agreed but maybe we can leave some of these code  optimizations to when we have a VIT model inside netket ? I think for a tutorial this is already OK? (for example the tutorial on the backflow I wrote was not meant to be exhaustive and has maybe worse limitations than this one, I think !)

PhilipVinc commented:
This is a tutorial, and the code needs to be clear. 
I do not think the current version of this code to be particularly clear. 

What I am asking is to make the tutorial clear, not more performant. Asserts and such are ways to make it more readable.
The changes I'm asking would take under one hour to implement and make the code considerably easier to read. 

The issues I'm raising stem from having seen several people struggle and complain that Luciano's model is harder to understand because the name of those variables do not make the code particularly clear. 
As 90% of people will end up copy pasting those snippets, I do not want to encourage *bad* code.

gcarleo commented:
that's the thing, they are not supposed to use code snippets fron here, they are supposed to use the netket model. I understand you want to discourage bad (how bad though ?) code but the side effect is we discourage people from contributing code/ writing tutorials

gcarleo commented:
the tutorial, by definition, is a piece of code that exemplifies usage in a very specific case, and that is not strictly guaranteed to work in other cases beyond the scope of the tutorial itself. this is the practice in many libraries including flax etc. for performant and Universal code one needs to use a library of models, not snippets of a model written for a simple tutorial

PhilipVinc commented:
> the tutorial, by definition, is a piece of code that exemplifies usage in a very specific case

We have no such tutorials on netket (or almost none). besides, they are most likely not useful because NetKet does not work out of the box and you need to customise several things. The single largest complaint  of **everybody** is that NetKet is poorly documented. The tutorials are a way to **docuemnt** how netket (and related things) work. Writing a 'tutorial' that cannot be generalised to reasonable cases is not useful to most people.  

I think the [back flow tutorial](https://netket.readthedocs.io/en/latest/tutorials/lattice-fermions.html) is very cleanly written and the code is extensively documented. Names of arguments are well understandable and every non-obvious block of lines has several comments.

I've seen many students in many different places copy paste the networks from that tutorial as a starting point for building more complicated architectures. There is no back flow architecture in netket so that's the best they get. And besides, even if we had one, it could not do everything people want, so they would still use it as a starting point for their experimentation.

> that's the thing, they are not supposed to use code snippets fron here, they are supposed to use the netket model

I'm not sure what model you are discussing in here. 
This PR is a tutorial on writing and using a ViT wave-function. This is not included in NetKet.
The snippets in the [present PR](https://netket--2004.org.readthedocs.build/en/2004/tutorials/ViT-wave-function.html) are not as clear as the one in the back flow tutorial. 

Again, the best example is, how is one supposed to know that `L_eff` in this tutorial which corresponds in 1D to the number of patches, or the effective system size of the system seen by the encoder, which  makes sense, in 2D corresponds to the number of patches?
As this is a tutorial about the ViT wavefunction it seems reasonable that this and other comments be addressed.

llviteritti commented:
We have modified the names of the variables and we have changed the class ViT using `@nn.compact` in order to avoid the use of the number of patches as input of the architecture.
